HW2/P5/physics.pyx

```diff
@@ -69,31 +73,42 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
-
+    
+    grid_x = int(XY1[0] / grid_spacing)
+    grid_y = int(XY1[1] / grid_spacing)
+    ## Check the nearby three grids.
+    for x in range(max(0, grid_x - 3), min(Grid.shape[0], grid_x + 4)):
```

Comment:
Since you're only checking objects up to two spaces away, you could use max(0, grid_x - 2), and min(Grid.shape[0], grid_x + 3).